### DISCUSSION TITLE
generic constrains for enum

### DISCUSSION THREAD

**Discussion by ali50m** [dotnet/roslyn#69730] [General] [answered]
I think the generic constrain should not accept the **Yoo** format. Since you can define it, but can not use it as type. 

Could someone help to point out how to use this **Yoo** generic since I can define it. 

```csharp
public class Foo<T> where T : struct, Enum
{
    public T Value { get; set; }
}


public class Yoo<T> where T : class, Enum
{
    public T Value { get; set; }
}

public class Bar
{
    public Foo<Goo> Goo { get; set; } = new();

    // this line will cause compiling error!
    public Yoo<Goo> Yoo { get; set; } = new();
}

public enum Goo
{
    A,
    B,
}
```

![image](https://github.com/dotnet/roslyn/assets/9393831/dad433b4-ccca-426a-b07d-5961e57e2988)

**Comment by CyrusNajmabadi** [comment]
> Could someone help to point out how to use this Yoo generic since I can define it.

you can do `Yoo<System.Enum>`.  You'd also be able to use any other reference type subclass of System.Enum if any are added by the BCL in the future.

**Comment by ali50m** [comment]
@CyrusNajmabadi thanks for your reply. But I can`t define Foo<System.Enum> as type now. Maybe the Foo<T> should not have the constrains both for struct and Enum at the same time.

![image](https://github.com/dotnet/roslyn/assets/9393831/903d2e7f-cbbb-4b1d-9087-5498ce5f1e99)

**Reply by CyrusNajmabadi** [reply, marked as answer]
> thanks for your reply. But I can`t define Foo<System.Enum> as type now. 

Correct.  But you asked `how to use this Yoo generic`.  You can use the `Yoo` generic with the `System.Enum` type.  You can use the `Foo` generic with any actual `enum` type.

> Maybe the Foo should not have the constrains both for struct and Enum at the same time.

The `Foo` generic makes complete sense.  It is saying that to be instantiated the type must be a subtype of System.Enum, and that it must be a struct.   In .net today there are tons of types that satisfy this. Namely, every single `enum`.

The `Yoo` generic is odd, but still makes sense.  It is saying that to be instantiated the type must be a subtype of System.ENum (or System.Enum itself), and that it must be a class.  In .net *today* only one type that satisfies this (System.Enum itself).  Perhaps in the future there will be more.  For now though, it's probably not something anyone would want to do.

**Reply by zacharylayne** [reply]
> Perhaps in the future there will be more.

CONFIRMED: C# 13 introducing Enum classes.

**Reply by ali50m** [reply]
@zacharylayne OMG! Do you have the link?

**Reply by huoyaoyuan** [reply]
> CONFIRMED: C# 13 introducing Enum classes.

No. C# 13 doesn't contain any change related to enums.

Even if some concepts are going to happen in the future, it's still very unlikely to see any new class inheriting from System.Enum, because it inherits from System.ValueType, which blocks non-special classes inheriting from it.

**Comment by ali50m** [comment]
I found if I remove Enum from Foo constrains, then `Foo<int>` will possiable now.

**Reply by CyrusNajmabadi** [reply]
Correct.  Because now you're not constraining to subtypes of System.Enum.
